### PR TITLE
Create explicit classes for node construction circular references

### PR DESCRIPTION
**Pull request by thecoop**
Create classes for RerouteService and RepositoriesService circular references. This hides the fact that there are circular references from the classes themselves - its `NodeConstruction`s responsibility to make sure services are called with things they can use, not for services to work round implementation details outside of themselves

As part of this, I've pulled out an interface for plugins to access repositories, allowing delegation during node construction

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/plugins/Plugin.java [hunk 1]
```diff
@@ -125,11 +124,10 @@ public interface PluginServices {
...
 
         /**
-         * A supplier for the service that manages snapshot repositories.
-         * This will return null when {@link #createComponents(PluginServices)} is called,
-         * but will return the repositories service once the node is initialized.
+         * An object to access repositories.
+         * Note that this will not fully work until node has been initialized.
          */
-        Supplier<RepositoriesService> repositoriesServiceSupplier();
+        Repositories repositories();
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/plugins/Plugin.java:130]: Alternatively, is there some reason `RepositoriesService` needs to be created after plugin components are obtained? Can it be moved to before?

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Review by rjernst — COMMENTED** [review]
Instead of propagating the circular reference, can we fix the issue? It seems to be a more explicit version of the guice proxy objects. I would rather spend the time to detangle the dependencies than add more solidity to them (new proxy interfaces + repositories interface).

**Comment by thecoop** [comment]
the createComponents move #103270 broke - reserved state handlers is needed by `ActionModule`, which then goes `NetworkModule` -> `TransportService` -> `RepositoriesModule`, but plugins need `createComponents` called before reserved state handlers are built

**Comment by thecoop** [comment]
We're not going to do this for now, we don't want to make it easier to create circular references

**Pull request closed without merging**